File: Matury CKE/2023 Przykladowy Arkusz/Programowanie/Python/zad1_3.py
# dane=[i.strip() for i in open('../../Dane_2203/szachy_przyklad.txt')]
dane=[i.strip() for i in open('../../Dane_2203/szachy.txt')]

i=0
plansze=[[]]
for row in dane:
    if row == '':
        i+=1
        plansze.append([])
    else:
        plansze[i].append(row)

# Cke dodało dwie nowe linie na końcu a to mogło popsuć program :o
if plansze[-1] == []:
    plansze.pop(len(plansze)-1)

szach_bialy=0
szach_czarny=0

def czy_szachuje(k, w, string):
    bialy = w.isupper()
    k_idx = string.index(k)
    if k_idx>0 and k_idx<len(string)-1:
        if string[k_idx-1] or string[k_idx+1] == w:
            return [True, bialy]
    elif k_idx == 0:
        if string[k_idx+1]==w:
            return [True, bialy]
    elif k_idx == len(string) - 1:
        if string[k_idx-1]==w:
            return [True, bialy]
    return [False, bialy]


for plansza in plansze:
    for i in range(len(plansza)):
        ten_wiersz=''
        ta_kolumna=''
        for j in range(len(plansza)):
            ten_wiersz+=plansza[i][j]
            ta_kolumna+=plansza[j][i]
        ten_wiersz=ten_wiersz.replace('.','')
        ta_kolumna=ta_kolumna.replace('.','')

        teksty=[ten_wiersz, ta_kolumna]
        bierki_cases=[['k', 'W'], ['K', 'w']]

        for tekst in teksty:
            for bierki in bierki_cases:
                if bierki[0] in tekst and bierki[1] in tekst:
                    szachuje = czy_szachuje(bierki[0], bierki[1], tekst)
                    if(szachuje[0]):
                        if(szachuje[1]):
                            szach_bialy+=1
                        else:
                            szach_czarny+=1

print(f'1.3. {szach_bialy} {szach_czarny}')

File: Matury CKE/2023 Przykladowy Arkusz/Programowanie/C++/zad1_3.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>

bool czy_szachuje(std::string bierki, char w, char k)
{
    int wCount = std::count(bierki.begin(), bierki.end(), w);
    int kCount = std::count(bierki.begin(), bierki.end(), k);
    if(wCount == 0 || kCount == 0)return false;

    int wIndex = bierki.find(w);

    if(wIndex == 0){
        if(bierki[1] == k){
            return true;
        }
    }else if(wIndex == bierki.size()-1){
        if(bierki[bierki.size()-2] == k){
            return true;
        }
    }else{
        if(bierki[wIndex-1] == k){
            return true;
        }
        if(bierki[wIndex+1] == k){
            return true;
        }
    }

    return false;
}

int main()
{
    // std::ifstream dane("C:/Matury-Informatyka/Matury CKE/2023 Przykladowy Arkusz/Dane_2203/szachy_przyklad.txt");
    std::ifstream dane("C:/Matury-Informatyka/Matury CKE/2023 Przykladowy Arkusz/Dane_2203/szachy.txt");

    std::vector<std::vector<std::string>> szachownice;

    std::string tempWiersz;
    std::vector<std::string> tempSzachownica;
    int s_i=0;
    while(dane >> tempWiersz){
        tempSzachownica.push_back(tempWiersz);

        s_i++;
        if(s_i==8){
            s_i=0;
            szachownice.push_back(tempSzachownica);
            tempSzachownica.clear();
        }
    }

    int szachujeBialyCount = 0;
    int szachujeCzarnyCount = 0;

    for(auto szachownica: szachownice){
        std::string szachuje = "";
        for(int i = 0; i<8; i++){
            std::string wiersz = "";
            std::string kolumna = "";

            for(int j = 0; j<8; j++){
                wiersz+=szachownica[i][j];
                kolumna+=szachownica[j][i];
            }

            wiersz.erase(std::remove(wiersz.begin(), wiersz.end(), '.'), wiersz.end());
            kolumna.erase(std::remove(kolumna.begin(), kolumna.end(), '.'), kolumna.end());

            std::vector<std::string> wierszKolumnaCases = {wiersz, kolumna};

            std::vector<std::vector<char>> bierkiCases = { {'W', 'k'}, {'w', 'K'} };


            for(auto wierszKolumna: wierszKolumnaCases){
                for(auto bierkaCase: bierkiCases){
                    bool szachuje = czy_szachuje(wierszKolumna, bierkaCase[0], bierkaCase[1]);
                    if(szachuje){
                        bool czyBialy = std::isupper(bierkaCase[0]);

                        if(czyBialy){
                            szachujeBialyCount++;
                        }else{
                            szachujeCzarnyCount++;
                        }

                        break;
                    }
                }
            }
        }
    }

    std::cout << "1.3. " << szachujeBialyCount << " " << szachujeCzarnyCount;

    return 0;
}
